### ISSUE TITLE
systemd-homed: increase LUKS sector size to 4kB

### ISSUE THREAD

**Issue by jre21** [systemd/systemd#24382]
### Component

homectl, systemd-homed

### Is your feature request related to a problem? Please describe

Systemd-homed's LUKS storage backend currently hardcodes the LUKS sector size to [512 bytes](https://github.com/systemd/systemd/blob/20e72c2bd42ddd2f39f3f4ac2384c4fecd4ff444/src/home/homework-luks.c#L1763).  This results in a significant performance cost for NVMe drives.  From my own testing, increasing this value to 4kB would increase I/O performance by roughly 50%.

To test the likely performance impact of this change, I created two loopback files on an unencrypted ext4 filesystem (note that setting the loopback device's block size causes cryptsetup to increase its own default sector size to match):

```
# fallocate -l 4G smallBlock
# losetup -f --show smallBlock
/dev/loop1
# echo '' | cryptsetup luksFormat -q /dev/loop1
# cryptsetup open /dev/loop1 smallBlock

# fallocate -l 4G largeBlock
# losetup -f -b 4k --show largeBlock
/dev/loop2
# echo '' | cryptsetup luksFormat -q /dev/loop2
# cryptsetup open /dev/loop2 largeBlock
```

I then used [fio](https://fio.readthedocs.io/en/latest/fio_doc.html) to configure six jobs testing different I/O patterns:
- smallRand: random 4k reads and writes against /dev/mapper/smallBlock
- largeRand: random 4k reads and writes against /dev/mapper/largeBlock
- smallRead: sequental 1M reads against /dev/mapper/smallBlock
- largeRead: sequental 1M reads against /dev/mapper/largeBlock
- smallWrite: sequental 1M writes against /dev/mapper/smallBlock
- largeWrite: sequental 1M writes against /dev/mapper/largeBlock

Each test was set to run for ten seconds, and I've included the most relevant output lines:

```
# fio smallRand.conf largeRand.conf smallRead.conf largeRead.conf smallWrite.conf largeWrite.conf
smallRand: (groupid=0, jobs=1): err= 0: pid=6985: Sun Aug 21 18:21:06 2022
  read: IOPS=30.3k, BW=118MiB/s (124MB/s)(1184MiB/10001msec)
  write: IOPS=30.3k, BW=118MiB/s (124MB/s)(1184MiB/10001msec); 0 zone resets
largeRand: (groupid=1, jobs=1): err= 0: pid=6991: Sun Aug 21 18:21:06 2022
  read: IOPS=46.2k, BW=180MiB/s (189MB/s)(1805MiB/10001msec)
  write: IOPS=46.3k, BW=181MiB/s (190MB/s)(1810MiB/10001msec); 0 zone resets
smallRead: (groupid=2, jobs=1): err= 0: pid=6993: Sun Aug 21 18:21:06 2022
  read: IOPS=1405, BW=1406MiB/s (1474MB/s)(13.7GiB/10001msec)
largeRead: (groupid=3, jobs=1): err= 0: pid=6996: Sun Aug 21 18:21:06 2022
  read: IOPS=2027, BW=2027MiB/s (2126MB/s)(19.8GiB/10001msec)
smallWrite: (groupid=4, jobs=1): err= 0: pid=6997: Sun Aug 21 18:21:06 2022
  write: IOPS=1002, BW=1002MiB/s (1051MB/s)(9.79GiB/10001msec); 0 zone resets
largeWrite: (groupid=5, jobs=1): err= 0: pid=7005: Sun Aug 21 18:21:06 2022
  write: IOPS=1540, BW=1540MiB/s (1615MB/s)(15.0GiB/10001msec); 0 zone resets
```

As an aside, cryptsetup's [man page](https://man.archlinux.org/man/core/cryptsetup/cryptsetup-luksFormat.8.en) has the following to say on the topic:

> `--sector-size bytes`
> Set sector size for use with disk encryption. It must be power of two and in range 512 - 4096 bytes. This option is available only with LUKS2 format.
>
> For LUKS2 devices it’s established based on parameters provided by underlying data device. For native 4K block devices it’s 4096 bytes. For 4K/512e (4K physical sector size with 512 bytes
emulation) it’s 4096 bytes. For drives reporting only 512 bytes block size it remains 512 bytes. If data device is regular file put in filesystem it’s 4096 bytes.
>
> Note that if sector size is higher than underlying device hardware sector and there is not integrity protection that uses data journal, using this option can increase risk on incomplete se
ctor writes during a power fail.
>
> If used together with --integrity option and dm-integrity journal, the atomicity of writes is guaranteed in all cases (but it cost write performance - data has to be written twice).
>
> Increasing sector size from 512 bytes to 4096 bytes can provide better performance on most of the modern storage devices and also with some hw encryption accelerators.

### Describe the solution you'd like

The main possibilities that occur to me:

1. Update the hardcoded value to from `512U` `4096U`
2. Set the sector size based on some heuristic (most likely the block size of the filesystem underneath the loopback file)
3. Allow setting the sector size through a homectl flag, possibly in combination with changing the default to one of the above

### The systemd version you checked that didn't have the feature you are asking for

251.4

**Comment by jre21** [comment]
Given how many different functions in homework-luks.c make byte/sector conversions, I presume this feature would also require code which detects the sector size of existing volumes to ensure the conversions use the correct ratio.

**Comment by AidanGG** [comment]
Just to add, mkfs.btrfs appears to detect the appropriate sector size, so there shouldn't need to be any changes to the execlp'd mkfs.btrfs call to make use of 4K sectors if it is exposed to by the LUKS volume.

**Comment by runiq** [comment]
Triage: Can this be closed since #24815 is implemented?
